sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc

```diff
@@ -0,0 +1,378 @@
+  friend marray<ComplexDataT, NumElements> operator op(const marray<ComplexDataT, NumElements> &lhs, const ComplexDataT &rhs) = delete;
+
+  // OP is: &=, |=, ^=
+  friend marray<ComplexDataT, NumElements> &operator op(marray<ComplexDataT, NumElements> &lhs, const marray<ComplexDataT, NumElements> &rhs) = delete;
+  friend marray<ComplexDataT, NumElements> &operator op(marray<ComplexDataT, NumElements> &lhs, const ComplexDataT &rhs) = delete;
+  friend marray<ComplexDataT, NumElements> &operator op(ComplexDataT &lhs, const marray<ComplexDataT, NumElements> &rhs) = delete;
+
+  // OP is: &&, ||
+  friend marray<bool, NumElements> operator op(const marray<ComplexDataT, NumElements> &lhs, const marray<ComplexDataT, NumElements> &rhs) = delete;
+  friend marray<bool, NumElements> operator op(const marray<ComplexDataT, NumElements> &lhs, const ComplexDataT &rhs) = delete;
+  friend marray<bool, NumElements> operator op(const ComplexDataT &lhs, const marray<ComplexDataT, NumElements> &rhs) = delete;
+
+  // OP is: <<, >>
+  friend marray<ComplexDataT, NumElements> operator op(const marray<ComplexDataT, NumElements> &lhs, const marray<ComplexDataT, NumElements> &rhs) = delete;
+  friend marray<ComplexDataT, NumElements> operator op(const marray<ComplexDataT, NumElements> &lhs, const ComplexDataT &rhs) = delete;
+  friend marray<ComplexDataT, NumElements> operator op(const ComplexDataT &lhs, const marray<ComplexDataT, NumElements> &rhs) = delete;
+
+  // OP is: <<=, >>=
+  friend marray<ComplexDataT, NumElements> &operator op(marray<ComplexDataT, NumElements> &lhs, const marray<ComplexDataT, NumElements> &rhs) = delete;
+  friend marray<ComplexDataT, NumElements> &operator op(marray<ComplexDataT, NumElements> &lhs, const ComplexDataT &rhs) = delete;
+
+  // OP is: ==, !=
+  friend marray<bool, NumElements> operator op(const marray<ComplexDataT, NumElements> &lhs, const marray<ComplexDataT, NumElements> &rhs);
+  friend marray<bool, NumElements> operator op(const marray<ComplexDataT, NumElements> &lhs, const ComplexDataT &rhs);
+  friend marray<bool, NumElements> operator op(const ComplexDataT &lhs, const marray<ComplexDataT, NumElements> &rhs);
+
+  // OP is: <, >, <=, >=
+  friend marray<bool, NumElements> operator op(const marray<ComplexDataT, NumElements> &lhs, const marray<ComplexDataT, NumElements> &rhs) = delete;
+  friend marray<bool, NumElements> operator op(const marray<ComplexDataT, NumElements> &lhs, const ComplexDataT &rhs) = delete;
+  friend marray<bool, NumElements> operator op(const ComplexDataT &lhs, const marray<ComplexDataT, NumElements> &rhs) = delete;
+
+  friend marray<ComplexDataT, NumElements> operator~(const marray<ComplexDataT, NumElements> &v) = delete;
+
+  friend marray<bool, NumElements> operator!(const marray<ComplexDataT, NumElements> &v) = delete;
+};
+
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+```
```

Comment:
Actually, I wonder why we want to add new member functions to `marray` at all.  Rather than doing this, let's add two new free functions like:

```
template<typename T, size_t NumElements>
marray<T, NumElements> real(const marray<complex<T>, NumElements> &z);

template<typename T, size_t NumElements>
marray<T, NumElements> imag(const marray<complex<T>, NumElements> &z);
```

This is consistent with standard C++, which has free functions `std::real(const std::complex<T> &z)` and `std::imag(const std::complex<T> &z)`.  There is no "marray" in standard C++, so there is no precedent for extending it with extra member functions.

Reply:
I understand your point and here's a little bit of context.

The reason for the `real` and `imag` being members of the `marray` specialization is the conclusion of a discussion that happened within this PR https://github.com/intel/llvm/pull/6550.

The goal was to have equivalent members within `marray`, so this struct could be abstract and you will apply the same member on the value that you are working on as if this was a `complex`.

Because `complex` has a `real` and `imag` member, `marray` has a `real` and `imag` member.

Reply:
There's a lot of discussion in #6550.  Can you summarize the reason for adding these member functions to `marray`?  Much of the discussion I see there is arguing to keep the API simple in this extension.  It seems to me that the API is simpler if we do not add new member functions to `marray`.

> The goal was to have equivalent members within `marray`, so this struct could be abstract and you will apply the same member on the value that you are working on as if this was a `complex`.

I'm not quite sure what you are arguing here.  Do you want to allow generic code that operates on either an `marray` of `complex` or on a single `complex`?  I'd recommend adding free functions `real` and `imag` functions that operate on either like:

```
template<typename T>
T real(const complex<T> &z);

template<typename T>
T imag(const complex<T> &z);

template<typename T, size_t NumElements>
marray<T, NumElements> real(const marray<complex<T>, NumElements> &z);

template<typename T, size_t NumElements>
marray<T, NumElements> imag(const marray<complex<T>, NumElements> &z);
```

This way, the generic code can use the free functions, not the member functions.

Reply:
That was the goal yes "allow generic code that operates on either an marray of complex or on a single complex".

Alright, So this would be the real/imag API:

```
complex.real();
complex.imag();
real(complex)
imag(complex);

// newest member real and imag are being deleted from the marray specialization

real(marray<complex>);
imag(marray<complex>);
```

Reply:
right!